File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/activity_intensity/activity_intensity_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/activity_intensity/activity_intensity_type_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';
import 'package:meta/meta.dart';

/// Extension to convert [ActivityIntensityRecord] to
/// [ActivityIntensityRecordDto].
@sinceV3_2_0
@internal
extension ActivityIntensityRecordToDto on ActivityIntensityRecord {
  ActivityIntensityRecordDto toDto() {
    return ActivityIntensityRecordDto(
      id: id.value,
      metadata: metadata.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      startZoneOffsetSeconds: startTime.timeZoneOffset.inSeconds,
      endZoneOffsetSeconds: endTime.timeZoneOffset.inSeconds,
      activityIntensityType: activityIntensityType.toDto(),
      title: title,
      notes: notes,
    );
  }
}

/// Extension to convert [ActivityIntensityRecordDto] to
/// [ActivityIntensityRecord].
@sinceV3_2_0
@internal
extension ActivityIntensityRecordDtoToDomain on ActivityIntensityRecordDto {
  ActivityIntensityRecord toDomain() {
    return ActivityIntensityRecord.internal(
      id: id?.toDomain() ?? HealthRecordId.none,
      metadata: metadata.toDomain(),
      startTime: DateTime.fromMillisecondsSinceEpoch(
        startTime,
        isUtc: true,
      ).toLocal(),
      endTime: DateTime.fromMillisecondsSinceEpoch(
        endTime,
        isUtc: true,
      ).toLocal(),
      activityIntensityType: activityIntensityType.toDomain(),
      title: title,
      notes: notes,
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/ActivityIntensityRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.ActivityIntensityRecord
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.ActivityIntensityRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.ActivityIntensityTypeDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [ActivityIntensityRecord] object to a [ActivityIntensityRecordDto].
 */
internal fun ActivityIntensityRecord.toDto(): ActivityIntensityRecordDto {
    val activityIntensityTypeDto = when (this.activityIntensityType) {
        ActivityIntensityRecord.ACTIVITY_INTENSITY_TYPE_MODERATE ->
            ActivityIntensityTypeDto.MODERATE

        ActivityIntensityRecord.ACTIVITY_INTENSITY_TYPE_VIGOROUS ->
            ActivityIntensityTypeDto.VIGOROUS

        else -> throw IllegalArgumentException(
            "Unsupported activity intensity type: ${this.activityIntensityType}",
        )
    }

    return ActivityIntensityRecordDto(
        id = metadata.id,
        startTime = startTime.toEpochMilli(),
        endTime = endTime.toEpochMilli(),
        startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
        endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
        metadata = metadata.toDto(),
        activityIntensityType = activityIntensityTypeDto,
        title = null,
        notes = null,
    )
}

/**
 * Converts a [ActivityIntensityRecordDto] to a Health Connect [ActivityIntensityRecord] object.
 */
internal fun ActivityIntensityRecordDto.toHealthConnect(): ActivityIntensityRecord {
    val activityIntensityType = when (this.activityIntensityType) {
        ActivityIntensityTypeDto.MODERATE ->
            ActivityIntensityRecord.ACTIVITY_INTENSITY_TYPE_MODERATE

        ActivityIntensityTypeDto.VIGOROUS ->
            ActivityIntensityRecord.ACTIVITY_INTENSITY_TYPE_VIGOROUS
    }

    return ActivityIntensityRecord(
        activityIntensityType = activityIntensityType,
        startTime = Instant.ofEpochMilli(startTime),
        endTime = Instant.ofEpochMilli(endTime),
        startZoneOffset = startZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        endZoneOffset = endZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )
}
